crates/binjs_io/src/context/prelude.rs

```diff
@@ -0,0 +1,214 @@
+    IdentifierNameLen,
+
+    /// The sequence of length of strings in `LiteralStringLen`.
+    LiteralStringLen,
+}
+impl Section {
+    /// Return the representation as a 4 bytes name.
+    pub fn as_bytes(&self) -> &'static Name {
+        match *self {
+            Section::Float => b"FLOA",
+            Section::UnsignedLong => b"ULON",
+            Section::ListLength => b"LIST",
+            Section::PropertyKeyContent => b"PROC",
+            Section::PropertyKeyLen => b"PROL",
+            Section::IdentifierNameContent => b"IDEC",
+            Section::IdentifierNameLen => b"IDEL",
+            Section::LiteralStringContent => b"STRC",
+            Section::LiteralStringLen => b"STRL",
+        }
+    }
+
+    /// Parse the 4 bytes name.
+    pub fn from_bytes(bytes: &Name) -> Option<Self> {
+        match bytes {
+            b"FLOA" => Some(Section::Float),
+            b"ULON" => Some(Section::UnsignedLong),
+            b"LIST" => Some(Section::ListLength),
+            b"PROC" => Some(Section::PropertyKeyContent),
+            b"PROL" => Some(Section::PropertyKeyLen),
+            b"IDEC" => Some(Section::IdentifierNameContent),
+            b"IDEL" => Some(Section::IdentifierNameLen),
+            b"STRC" => Some(Section::LiteralStringContent),
+            b"STRL" => Some(Section::LiteralStringLen),
+            _ => None,
+        }
+    }
+}
+
+/// Decode a prelude into an iterator of sections.
+pub struct PreludeDecoder<T>
+where
+    T: Read,
+{
+    reader: T,
+    name: Option<Name>,
+}
+impl<T> PreludeDecoder<T>
+where
+    T: Read,
+{
+    /// Create a `PreludeDecoder`.
+    ///
+    /// The `reader` is expected to contain the *content* of the prelude, *without*
+    /// its name `PREL`.
+    pub fn new(reader: T) -> Self {
+        PreludeDecoder { reader, name: None }
+    }
+
+    /// Finalize and recover the reader.
+    pub fn into_input(self) -> T {
+        self.reader
+    }
+
+    /// Return the latest name encountered by the `PreludeDecoder`.
+    ///
+    /// Note that if we have reached the end of the `PreludeDecoder`, this is the name of
+    /// the next section of the file.
+    pub fn name(&self) -> Option<&Name> {
+        self.name.as_ref()
+    }
+
+    /// Read the next section, update `self.name()`.
+    fn read_stream(&mut self) -> Result<Option<(Section, Vec<u8>)>, TokenReaderError> {
+        // Read section name.
+        let mut name = [0; 4];
+        self.reader
+            .read_exact(&mut name)
+            .map_err(TokenReaderError::ReadError)?;
+        let section = Section::from_bytes(&name);
+        self.name = Some(name);
+        let section = match section {
+            None => {
+                return Ok(None);
+            }
+            Some(section) => section,
+        };
+
+        // Read compression format.
+        let mut name = [0; 4];
+        self.reader
+            .read_exact(&mut name)
+            .map_err(TokenReaderError::ReadError)?;
```

Comment:
not about this code but the format itself,
is there any reason that we have compression format specifier in each stream?
there can be more than 10 streams in single file, and I don't see any reason to use different compression for each stream.
(at least for streams inside prelude/content)
are you planning to change it in the future, to use single compression type for prelude/content/main for each, or something along that line?

Reply:
The original reason was that I have witnessed cases in which Brotli made some streams quite larger for small files. I'm not 100% sure it's a good idea, though.

Reply:
Maybe we'll get rid of this in a future version?

Reply:
oh, indeed.
encoder can choose uncompressed format when brotli "compressed" stream becomes larger than original.
let's keep it as is for now :)